Source/WebCore/loader/ResourceMonitorThrottler.cpp

```diff
@@ -27,42 +27,80 @@
 #include "ResourceMonitorThrottler.h"
 
 #include "Logging.h"
-#include <wtf/CryptographicallyRandomNumber.h>
+#include "ResourceMonitorPersistence.h"
+#include <wtf/MainThread.h>
 #include <wtf/Seconds.h>
 #include <wtf/StdLibExtras.h>
-#include <wtf/text/StringHash.h>
+#include <wtf/text/MakeString.h>
 
 #if ENABLE(CONTENT_EXTENSIONS)
 
 #define RESOURCEMONITOR_RELEASE_LOG(fmt, ...) RELEASE_LOG(ResourceLoading, "%p - ResourceMonitorThrottler::" fmt, this, ##__VA_ARGS__)
 
 namespace WebCore {
 
-static constexpr size_t defaultThrottleAccessCount = 5;
-static constexpr Seconds defaultThrottleDuration = 24_h;
-static constexpr size_t defaultMaxHosts = 100;
-
 ResourceMonitorThrottler::ResourceMonitorThrottler()
     : ResourceMonitorThrottler(defaultThrottleAccessCount, defaultThrottleDuration, defaultMaxHosts)
 {
 }
 
 ResourceMonitorThrottler::ResourceMonitorThrottler(size_t count, Seconds duration, size_t maxHosts)
+    : ResourceMonitorThrottler(SQLiteDatabase::inMemoryPath(), count, duration, maxHosts)
+{
+}
+
+ResourceMonitorThrottler::ResourceMonitorThrottler(const String& databasePath)
+    : ResourceMonitorThrottler(databasePath, defaultThrottleAccessCount, defaultThrottleDuration, defaultMaxHosts)
+{
+}
+
+ResourceMonitorThrottler::ResourceMonitorThrottler(const String& databasePath, size_t count, Seconds duration, size_t maxHosts)
     : m_config { count, duration, maxHosts }
+    , m_workQueue { WorkQueue::create("ResourceMonitorThrottler Work Queue"_s) }
 {
     ASSERT(maxHosts >= 1);
     RESOURCEMONITOR_RELEASE_LOG("Initialized with count: %zu, duration: %.f, maxHosts: %zu", count, duration.value(), maxHosts);
+
+    protectedWorkQueue()->dispatch([this, path = crossThreadCopy(databasePath)] mutable {
```

Comment:
Capturing `this` without protection here is unsafe. Nothing guarantees `this` will still be alive when the lambda runs.

call `relaxAdoptionRequirement();` in the constructor if needed to make this happen.

Reply:
Ah, right.